### PR TITLE
P#60741 / P#52107 Update allow sorting by all field + P#55975 improve sorting controls

### LINKED ISSUE #152: P#52107 Allow sorting by all fields : gaps in the implementation
@tung-le-esg 
We found some bugs in the mentioned project. Could you please analyse?

1.) If "Sort by user selection" is unchecked, the old dropdown with fixed field selection is shown for "sort by", not the required new text field with selection of all fields.
2.) The text search does not find all active fields, e.g. year of construction (Baujahr) or floor (Etage) are missing.
3.) The order of the fields under "Popular" is not as in the specification screenshot. The field "Number of bathrooms" is missing in "Popular".
4.) In my case there are only 3 fields under "All", but there are many more fields active. All fields should be displayed under "All" that are not under "Popular".
5.) In order for the default sort to be current, the view must be reloaded first, if sort by user selection is on. Default sorting should be updated directly after "Save changes", if you have made changes in "Sort by".

Feedback:
1.) fixed
 2.) not fixed 
3.) Sorting not fixed. "Number of bathrooms" is there.
4.) not fixed 
5.) not fixed. "Standard sort" option is broken. No values in the dropdown.

### PR DESCRIPTION
**Pull request by hungnc89**
fix #152

### CODE REVIEW

**Review by jayay — APPROVED** [review]

**Comment by jayay** [comment]
@hungnc89 @kellyfa12 could you please check the tests?

**Comment by hungnc89** [comment]
> Thank you! Would you think it'd be possible to cover the new lines even better?

I will update more test

**Review by jayay — APPROVED** [review]

**Comment by michaelarnold-dev** [comment]
@hungnc89 the testers identified some issues Test with version 2.17.0-3-gff1432c5 on https://wpptest.onofficeweb.com/:

1.) If "Sort by user selection" is unchecked, the old dropdown with fixed field selection is shown for "sort by", not the required new text field with selection of all fields.
2.) The text search does not find all active fields, e.g. year of construction (Baujahr) or floor (Etage) are missing.
3.) The order of the fields under "Popular" is not as in the specification screenshot. The field "Number of bathrooms" is missing in "Popular".
4.) In my case there are only 3 fields under "All", but there are many more fields active. All fields should be displayed under "All" that are not under "Popular".
5.) In order for the default sort to be current, the view must be reloaded first, if sort by user selection is on. Default sorting should be updated directly after "Save changes", if you have made changes in "Sort by". (We already deleted the browser cache but it did not help)

**Comment by jayay** [comment]
@michaelarnold-dev commit ff1432c5 from the test is from Aug 20, a newer test zip would be more representative.

**Review by jayay — APPROVED** [review]

**Review by jayay — APPROVED** [review]

File: plugin/Model/FormModelBuilder/FormModelBuilder.php [hunk 1]
```diff
@@ -88,13 +88,17 @@ protected function readFieldnames($module, bool $withInactive = false)
 		try {
 			$this->_pFieldnames->loadLanguage();
 			$fieldnames = $this->_pFieldnames->getFieldList($module);
-
+			$typeSupport = ['integer', 'float', 'varchar', 'text', 'date', 'datetime', 'boolean'];
```
Comment (jayay) [inline plugin/Model/FormModelBuilder/FormModelBuilder.php:91]: This looks legit, but is it needed here? Feels like it should be added for both Fieldnames and FieldsCollectionBuilders, but in a separate project

**Comment by michaelarnold-dev** [comment]
@hungnc89 Testing feedback: the errors are all still there. Could you please investigate again?

Version 2.21.2-12-g308568bf installed on https://1.wptest.onofficeweb.com/

**Comment by tang-hien-egs** [comment]
> @hungnc89 Testing feedback: the errors are all still there. Could you please investigate again?
> 
> Version 2.21.2-12-g308568bf installed on https://1.wptest.onofficeweb.com/

@michaelarnold-dev , the errors fixed, i recorded in video link, can you check it ? , the link : https://files.fm/f/7cgwtdeve

**Comment by michaelarnold-dev** [comment]
@tang-hien-egs 1.) If "Sort by user selection" is unchecked, the old dropdown with fixed field selection is shown for "sort by", not the required new text field with selection of all fields.
0:08 in the Video.

2.) The text search does not find all active fields, e.g. year of construction (Baujahr) or floor (Etage) are missing.
0:27 in the Video.

3.) The order of the fields under "Popular" is not as in the specification screenshot. The field "Number of bathrooms" is missing in "Popular".
0:27 in theVideo.

4.) In my case there are only 3 fields under "All", but there are many more fields active. All fields should be displayed under "All" that are not under "Popular".
cannot be seen in video

5.) In order for the default sort to be current, the view must be reloaded first, if sort by user selection is on. Default sorting should be updated directly after "Save changes", if you have made changes in "Sort by".
0:16 in the  Video.

**Comment by michaelarnold-dev** [comment]
@tang-hien-egs @nglelinh Any updates here?

**Comment by tang-hien-egs** [comment]
> @tang-hien-egs @nglelinh Any updates here?

@michaelarnold-dev , i tested in localhost, Record in tested local, in video like link : https://files.fm/f/pekvanp9p, i think you ask your tester should get code latest

**Comment by michaelarnold-dev** [comment]
@tang-hien-egs According to our tester Robert Igelmund some bugs still exist. The errors 3-5 can be observed in that video. Could you please analyze?

https://user-images.githubusercontent.com/77264788/157241682-ac924a87-96f0-41a7-952e-a6548483998c.mp4

**Comment by michaelarnold-dev** [comment]
@tang-hien-egs  Point 2.), 3.) and 4.) are not fixed for me. Can you record a video that shows that the bugs are fixed? The sorting of the fields under "Popular".That all fields are visible under "All". That fields "Year of construction" and "Floor" are found

**Comment by LongTrong-exe** [comment]
@michaelarnold-dev Sorry for wasting your time, I will take this task from now.

**Review by jayay — DISMISSED** [review]

**Review by jayay — APPROVED** [review]

**Review by jmaas-onoffice — APPROVED** [review]

**Comment by LongTrong-exe** [comment]
@michaelarnold-dev 
I fixed both errors you said, but bug 1) doesn't seem to be mentioned in the requirements.
This's my test video:

https://user-images.githubusercontent.com/101325858/162416231-eb6a551c-cd1d-41f8-b27b-2b686caca5f0.mp4

**Comment by hungnc89** [comment]
- Tester tested
- Link video : https://files.fm/u/xcyvs5pgy#/view/4mtrq5ukg

**Review by jayay — DISMISSED** [review]

**Comment by LongTrong-exe** [comment]
@jmaas-onoffice Can you test this task?

**Review by jmaas-onoffice — APPROVED** [review]

**Review by jmaas-onoffice — APPROVED** [review]

**Review by jmaas-onoffice — CHANGES_REQUESTED** [review]
The requirements seem to not have been understandable enough. I will discuss this with Linh next Wednesday.

The issue:

We have two different types of sorting. A "simple" sort where you just choose how the estates are ordered and a "user" sort where you give the user a dropdown with which they can sort the estates. Which of the two modes is active is determined by the checkbox "Sort by user selection."

It seems that the "user" sort interface was copied to the "simple" sort. When the checkbox is not active, it should look like this:

grafik

But in [3.1-32-gf25a2d75](https://github.com/onOfficeGmbH/oo-wp-plugin/pull/109#issuecomment-1119470104) it looks like this:

grafik

I think it's easiest if I dicuss this with Linh next Wednesday. I think the requirements weren't very clear about this, so if we talk about it, we should be able to solve the problem. :)

**Comment by LongTrong-exe** [comment]
@jmaas-onoffice When I received this task, the requirements error were not mentioned in the error report. So I thought it was okay, and I just fixed the bug that was reported by your tester.
By the way, can you resend me requirement?

**Comment by LongTrong-exe** [comment]
> The requirements seem to not have been understandable enough. I will discuss this with Linh next Wednesday.
> 
> The issue:
> 
> We have two different types of sorting. A "simple" sort where you just choose how the estates are ordered and a "user" sort where you give the user a dropdown with which they can sort the estates. Which of the two modes is active is determined by the checkbox "Sort by user selection."
> 
> It seems that the "user" sort interface was copied to the "simple" sort. When the checkbox is not active, it should look like this:
> 
> grafik
> 
> But in [3.1-32-gf25a2d75](https://github.com/onOfficeGmbH/oo-wp-plugin/pull/109#issuecomment-1119470104) it looks like this:
> 
> grafik
> 
> I think it's easiest if I dicuss this with Linh next Wednesday. I think the requirements weren't very clear about this, so if we talk about it, we should be able to solve the problem. :)



> The requirements seem to not have been understandable enough. I will discuss this with Linh next Wednesday.
> 
> The issue:
> 
> We have two different types of sorting. A "simple" sort where you just choose how the estates are ordered and a "user" sort where you give the user a dropdown with which they can sort the estates. Which of the two modes is active is determined by the checkbox "Sort by user selection."
> 
> It seems that the "user" sort interface was copied to the "simple" sort. When the checkbox is not active, it should look like this:
> 
> grafik
> 
> But in [3.1-32-gf25a2d75](https://github.com/onOfficeGmbH/oo-wp-plugin/pull/109#issuecomment-1119470104) it looks like this:
> 
> grafik
> 
> I think it's easiest if I dicuss this with Linh next Wednesday. I think the requirements weren't very clear about this, so if we talk about it, we should be able to solve the problem. :)

@jmaas-onoffice I just checked this function on master brand and got what you mean. I think it's not difficult to understand.
If you have any problem with the requirements, please discuss with Linh.

**Comment by LongTrong-exe** [comment]
Is this okay?

https://user-images.githubusercontent.com/101325858/167771368-7722aa05-18cd-4aaa-bd67-422f7fbc15c8.mp4

**Comment by jmaas-onoffice** [comment]
@LongTrong-exe Is the code from [your video](https://github.com/onOfficeGmbH/oo-wp-plugin/pull/109#issuecomment-1123183525) in this PR? Because when I install [3.1-32-gf25a2d75](https://github.com/onOfficeGmbH/oo-wp-plugin/pull/109#issuecomment-1119470104) it doesn't work like in your video, as you can see on https://2.wptest.onofficeweb.com/.

**Comment by LongTrong-exe** [comment]
@jmaas-onoffice New code is still in my local PC, I will push it tomorrow. I need to test it more for sure.

**Comment by jmaas-onoffice** [comment]
Ah, ok, now I understand. Awesome, then just request a review when it's ready. Thanks! :)

**Review by jmaas-onoffice — APPROVED** [review]

**Review by jmaas-onoffice — CHANGES_REQUESTED** [review]
1. In the simple sort mode (not the user sort mode), you don't have access to all fields when sorting.
    grafik
2. You can currently select unsupported fields. The highlighted field is a separator:
    grafik
    Only valid field types should be selectable. Which fields are valid is defined by #168.

**Comment by LongTrong-exe** [comment]
@jmaas-onoffice I'm filtering the field by these types:
image
So I think field "---------" you are talking about has a problem with its label, not its type

**Comment by jmaas-onoffice** [comment]
@LongTrong-exe Good point. Unfortunately it seems that I can actually select a separator:

grafik

I have updated https://1.wptest.onofficeweb.com/ and here is a screenshot of the configured fields:

grafik

I shouldn't be able to select for example "BLACK HINT", but I can:

grafik

So it seems that the fields are not filtered correctly. Could you fix it?

**Comment by jmaas-onoffice** [comment]
Also, should we merge this with #175, since they might conflict?

**Comment by LongTrong-exe** [comment]
@jmaas-onoffice This is weird, I think there is a problem with your API. This is the response of "BLACK HINT" field, as you can see, its type is "text", even though its type in onoffice is "Schwarzer Hinweis".
image

**Comment by jmaas-onoffice** [comment]
Yes, I had the same issue. You have to set the flag `realDataTypes` (https://apidoc.onoffice.de/actions/informationen-abfragen/feldkonfiguration/).

**Review by jmaas-onoffice — APPROVED** [review]

**Review by jmaas-onoffice — CHANGES_REQUESTED** [review]
1. When you activate the random sort, the checkbox for "Sort by user selection" gets deselected but the fields below are still the ones for user selection.

    https://user-images.githubusercontent.com/90763364/168553189-2e1828e0-b6ae-4a75-9536-4b598783c0d5.mp4

   When random sort is selected, the fields above should only be deactivated. The checkbox state should not be changed.

2. I'm not sure if this is in the original requirements, but in the user selection sort after adding criteria to "Sort by", I can't immediately select them in "Standard sort". I need to save the page, before I can select them.

   https://user-images.githubusercontent.com/90763364/168553527-f31ee0d4-a8b0-4748-aca1-40a4f1a4c2ba.mp4

   Could you make it so that when I add search criteria to "Sort by" that they are immediately selectable in "Standard sort"?

**Comment by LongTrong-exe** [comment]
@jmaas-onoffice Is this what you want? This is on my local PC.

https://user-images.githubusercontent.com/101325858/168558548-e27458d3-db11-4bdb-82af-eb7c2f5e6ff6.mp4

**Comment by LongTrong-exe** [comment]
@jmaas-onoffice Is this what you want? This is on my local PC.

https://user-images.githubusercontent.com/101325858/168563018-2e01fa48-a6ba-4702-ba9a-e431fe66ce18.mp4

**Review by jmaas-onoffice — APPROVED** [review]

**Review by jmaas-onoffice — CHANGES_REQUESTED** [review]
I've found a small issue when updating:

1. In version 3.1 make an estate list and set its ordering to "Random".
2. Update to the newest code.
3. The "Random" checkbox is correctly set, but the "Sort by User Selection" checkbox is not disabled.

    https://user-images.githubusercontent.com/90763364/168754262-640cfbd0-13cc-47df-90ff-3e3db11daf33.mp4

Apart from that it looks very good! I think we will be finally able to merge this soon!

**Comment by LongTrong-exe** [comment]
@jmaas-onoffice Thanks, you are such a great tester. I fixed it.

**Review by jmaas-onoffice — DISMISSED** [review]

**Comment by jmaas-onoffice** [comment]
@LongTrong-exe I forgot to ask in my last comment: This was not in the requirements, but for address lists you can still only select a fixed set of sorting criteria. How much time would it take to allow selecting any field for sorting address lists as well?

**Comment by LongTrong-exe** [comment]
> @LongTrong-exe I forgot to ask in my last comment: This was not in the requirements, but for address lists you can still only select a fixed set of sorting criteria. How much time would it take to allow selecting any field for sorting address lists as well?

@jmaas-onoffice I think it could take about three to five days, but you should create a new task for it.

**Comment by jmaas-onoffice** [comment]
@LongTrong-exe Ok, then let's leave it for now. Address lists are rarely used, so I don't want to put too much effort there.

Can we merge this with #175? I think it makes sense to release both together.

**Comment by LongTrong-exe** [comment]
@jmaas-onoffice I merged them, I think it looks good, can you test it?

**Review by jmaas-onoffice — APPROVED** [review]

**Review by jmaas-onoffice — CHANGES_REQUESTED** [review]
I found a small bug:

It seems that there is a problem with "Random sort". I think when the backend says "Random sort" is selected, it actually still has "Default sort" selected.

There are a few ways that this causes problems:

1. When I create a new estate list, the backend shows "Random sort". It should show select "Default sort". A few observations:
    - When I save this new list without changes (i.e. with "Random sort"), and then go to the page in the frontend with the shortcode and reload the page, the order never changes. If "Random sort" was active, the order of the estates should change on every page load.
    - I can set the sorting to "Default sort", change the settings and save, and they will be active. E.g. I can set "Sort order" to "Descending" and in the frontend the order flips. But when I then try to select "Random sort" and save, the order in the frontend is the same as before and it doesn't change when I reload the page.
2. If in v3.1 I have a list set to "Default sort" (i.e. "Sort by user selection" and "Random sort" are not checked) and then upgrade to the current code, the backend will show "Random sort" but the frontend will be correctly sorted with the settings from before the update. Also, when in the backend you switch to "Default sort", it shows the correct settings (e.g. if before the update I selected "Number of rooms" for "Sort by", it now also shows "Number of rooms" correctly).

Overall, I think that this is a GUI issue in the estate list settings.

**Comment by LongTrong-exe** [comment]
This's my test video:

https://user-images.githubusercontent.com/101325858/169272379-84359ae1-80d4-4dc9-95a6-fd2bad7549f7.mp4

**Comment by LongTrong-exe** [comment]
> 2. If in v3.1 I have a list set to "Default sort" (i.e. "Sort by user selection" and "Random sort" are not checked) and then upgrade to the current code, the backend will show "Random sort"

Thank you, but I can't reproduce (2.) "Default sort" list is still "Default sort" when I update to the new version.

**Comment by jmaas-onoffice** [comment]
Maybe you've fixed it. I think both problems have the same cause. I'll test it. 👍

**Review by jmaas-onoffice — APPROVED** [review]

**Review by jmaas-onoffice — CHANGES_REQUESTED** [review]
Unfortunately, I still have the second bug. Here is how I can reproduce it:

1. In v3.1 create a new estate list. Just add a name, do not change the settings.
    grafik
2. Update to [3.1-72-g09b8e217](https://github.com/onOfficeGmbH/oo-wp-plugin/pull/109/#issuecomment-1131584862).
3. Open the settings of the estate list. The sorting is set to "Random sort".
    grafik

If you still can't reproduce this, I will set up a test instance when one is free.

**Comment by LongTrong-exe** [comment]
@jmaas-onoffice yes, I understand what you say. There is a backward compatibility problem when changing from checkbox to select type. I'll fix it.

**Comment by LongTrong-exe** [comment]
https://user-images.githubusercontent.com/101325858/169446014-b8536a36-3b34-412d-a335-db75fdf27d39.mp4

**Review by jmaas-onoffice — APPROVED** [review]

**Pull request merged**